linkerd/stack/src/gate.rs

```diff
@@ -173,8 +173,15 @@ where
     type Future = S::Future;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        let permit = ready!(self.poll_acquire(cx));
-        ready!(self.inner.poll_ready(cx))?;
+        // If we have previously acquired a permit, move it out of the `Gate` in
+        // order to poll the inner service. If the inner service is no longer
+        // ready, we will drop our previously acquired permit.
+        let permit = match std::mem::replace(&mut self.permit, Poll::Pending) {
+            Poll::Ready(permit) => permit,
+            Poll::Pending => ready!(self.poll_acquire(cx)),
+        };
+
+        ready!(self.inner.poll_ready(cx)?);
```

Comment:
```suggestion
        if self.permit.is_ready() {
            return Poll::Ready(Ok(()));
        }
        let permit = ready!(self.poll_acquire(cx));
        ready!(self.inner.poll_ready(cx))?;
```

This seems like a simpler fix. We only ever store the permit when the inner service has polled ready. There's no need to poll it again, is there?

Reply:
That was actually what I had originally written, and it does also fix the problem. I thought that it was potentially desirable to continue polling the inner service in case its `poll_ready` implementation actually does work (e.g. if the gate is around a load balancer which needs to be polled to drive discovery). I'm not sure if this actually matters, though.

Reply:
Or: is there actually a situation where a service that has polled ready can later return pending? My understanding is that this violates the poll contract.

Reply:
You're correct that it should not return pending unless the readiness is consumed. I was just concerned about cases where `poll_ready` drives _work_ (e.g. polling discovery), but I think your simpler solution is fine here.

Reply:
I was wrong. Balancers rely on reconnect being able to go from Ready => Pending.